cou/apps/auxiliary.py

```diff
@@ -502,3 +508,156 @@ async def _verify_nova_compute(self, target: OpenStackRelease) -> None:
+        """
+        await self.model.wait_for_idle(
+            timeout=self.wait_timeout,
+            status="blocked",
+            apps=[self.name],
+            # Vault application will go into error status first then blocked status.
+            raise_on_error=False,
+        )
+
+        app_status = await self.model.get_application_status(app_name=self.name)
+        if not app_status.status.info == "Unit is sealed":
+            raise ApplicationError("Vault not in sealed status")
+        logger.debug("Application 'vault' in sealed status")
+
+    async def _get_vault_client(self, unit_name: str, cacert_file: Optional[str]) -> hvac.Client:
+        """Get vault client.
+
+        :param unit_name: vault unit name
+        :type unit_name: str
+        :param cacert_file: cacert file path
+        :type cacert_file: str
+        :return: hvac vault Client object
+        :rtype: hvac.Client
+        """
+        vault_url = await self._get_unit_api_url(unit_name)
+        client = hvac.Client(url=vault_url, verify=cacert_file)
+        return client
+
+    async def _unseal_vault(self) -> None:
+        """Unseal vault on every vault unit."""
+        # Stop progress_indicator because it will clear the unseal key input.
+        progress_indicator.stop()
+
+        for unit_name in self.units:
+            logger.debug("Start unseal %s", unit_name)
+            cacert_file = self._get_cacert_file()
+            client = await self._get_vault_client(unit_name, cacert_file)
+            while True:
+                status = client.sys.read_seal_status()
+                if not status["sealed"]:
+                    break
+                unseal_key = getpass.getpass("Unseal Key (will be hidden):")
+                if unseal_key:
+                    client.sys.submit_unseal_key(key=unseal_key)
+                del unseal_key
+
+            # Delete temporary ca file
+            if cacert_file:
+                os.remove(cacert_file)
+                logger.debug("Remove tempfile: %s", cacert_file)
+
+    def post_upgrade_steps(
+        self, target: OpenStackRelease, units: Optional[list[Unit]]
+    ) -> list[PostUpgradeStep]:
+        """Post Upgrade steps planning.
+
+        Wait until the application reaches the idle state and then check the target workload.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :param units: Units to generate post upgrade plan
+        :type units: Optional[list[Unit]]
+        :return: List of post upgrade steps.
+        :rtype: list[PostUpgradeStep]
+        """
+        upgrade_step = self._get_upgrade_charm_step(target=target)
+        steps = []
+
+        # Add unseal steps only if chaneel is changed.
+        if upgrade_step:
+            steps.extend(
+                [
+                    # Vault application should get into blocked and sealed status after upgrading.
```

Comment:
Are there cases where it will not become sealed?  For example, if there wasn't any available updates for vault and it doesn't upgrade and restart?

Reply:
The logic inside the charm is every time the channel upgrade, it will be sealed. I don't think the situation exists.